### PR TITLE
Reinstate the dependency injection engine

### PR DESCRIPTION
**Pull request by Konamiman**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

The dependency injection engine that [was introduced in 4.4.0](https://github.com/woocommerce/woocommerce/pull/26731) had to be [removed in 4.4.1](https://github.com/woocommerce/woocommerce/pull/27395) because it conflicted with a different version of the same package used by an extension.

This pull request reintroduces all the removed stuff but in order to avoid conflicts all the classes from the PHP League's Container package are re-namespaced, from the original `League\Container` to `Automattic\WooCommerce\Vendor\League\Container`. The re-namespacing is performed via a simple regex substitution in the code files, this approach works since the package code references its namespace in `use` and `namespace` statements only.

### How to test the changes in this Pull Request:

Run unit tests and do general smoke-testing.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Introduce (again) a dependency injection framework for the code in the src directory.

### CODE REVIEW

File: bin/package-update.sh
Comment (ObliviousHarmony) [inline bin/package-update.sh:3]: One of the caveats is that moving this function away also removes the documentation regarding what the color codes mean. I think it might be wise to leave it the way it is for clarity. Since there is only a single output statement in the other script how would you feel about outputting the color itself there?

File: bin/prefix-vendor-namespaces.sh
Comment (ObliviousHarmony) [inline bin/prefix-vendor-namespaces.sh:15]: What do you think of this instead?

```suggestion
find ./vendor/league/container -iname '*.php' -exec sed -i '.bak' -E -e "$REGEX" {} \;
find ./vendor/league/container -name "*.php.bak" -type f -delete
```

File: composer.json
Comment (ObliviousHarmony) [inline composer.json:48]: I'm not sure about this one, though, it does raise an interesting problem with PSR-4 namespaces and this method of renaming. I'm wondering if we should also correct the namespace in the `composer.json` for the dependency that way Composer knows what the namespace is when generating it?

**Review by ObliviousHarmony — COMMENTED** [review]
Thanks for tackling this @Konamiman! While I think perhaps this approach may work for fixing the DI container in the short-term, I do think we should consider a more complete solution.

Given our current build pipeline, I wonder if it might be better to perform this transformation in a Composer plugin prior to building the autoloader? This would ensure that the first time Composer executes it has the correct namespaces internally and would get around some of the problems we're going to encounter that require executing the autoloader twice.

**Comment by Konamiman** [comment]
> While I think perhaps this approach may work for fixing the DI container in the short-term, I do think we should consider a more complete solution.

Sure, that was the entire point from the beginning 🙂

> Given our current build pipeline, I wonder if it might be better to perform this transformation in a Composer plugin prior to building the autoloader? This would ensure that the first time Composer executes it has the correct namespaces internally and would get around some of the problems we're going to encounter that require executing the autoloader twice.

This would fall more into the "more complete solution" and I wouldn't do it right now as it defeats the point of the "short-term" stratgey. The `composer dump-autoload` stage is performed after the transformation anyway, so it doesn't seem to be an issue from the autoloading perspective.

File: bin/package-update.sh
Comment (Konamiman) [reply bin/package-update.sh:3]: I don't like the idea of duplicating code. What I could do is adding a comment saying that `output.sh` contains the colors reference, and a comment in that file itself explaining what it does and how to reuse it.

File: bin/prefix-vendor-namespaces.sh
Comment (Konamiman) [reply bin/prefix-vendor-namespaces.sh:15]: Sure, I can do this. 👍🏻

File: composer.json
Comment (Konamiman) [reply composer.json:48]: You mean changing `autoload` and `autoload-dev` in `vendor/league/container/composer.json`? Why not, if only for consistency (doesn't seem to be needed or am I missing something?)

Comment (ObliviousHarmony) [reply composer.json:48]: The problem I have with putting this PSR-4 mapping here is that it doesn't feel like something that belongs in the root `composer.json` since we're declaring a new namespace for code that already has a different namespace declaration elsewhere. Since we've renamed the namespace inside the package vendor files I believe we should do the same in the package's `composer.json`. 

For context, the next version of the jetpack-autoloader we utilize will support PSR-4 namespacing in development, which will use that declared namespace. If we don't change the package name there too then it will cause problems!

File: bin/package-update.sh
Comment (ObliviousHarmony) [reply bin/package-update.sh:4]: While I'd generally agree, I think the purpose of this `output` function is specific enough that it's fine to duplicate the code. To me this just makes the file more difficult to read if you're trying to tell what color is being output. If we're set on refactoring this function, perhaps we can offer more value and change the argument from an arbitrary integer to the color itself?

File: composer.json
Comment (ObliviousHarmony) [reply composer.json:48]: Just making sure you didn't miss this reply @Konamiman!

File: bin/package-update.sh
Comment (Konamiman) [reply bin/package-update.sh]: Ok, I have "de-deduplicated" the function. This is supposed to be a temporary thing anyway.

File: composer.json
Comment (Konamiman) [reply composer.json:48]: Sorry I think I wasn't clear enough. After you wrote that comment I updated the `prefix-vendor-namespaces.sh` script so that it also changes the namespace inside `vendor/league/container/composer.json`. This is what you meant by "also correct the namespace in the composer.json for the dependency", right?

**Review by ObliviousHarmony — DISMISSED** [review]
As a "get it working" solution I think this is fine but I'd love to see something more complete ASAP! I'm going to approve the PR pending resolution of `composer.json` conflicts and successful tests passing!

**Comment by Konamiman** [comment]
> As a "get it working" solution I think this is fine but I'd love to see something more complete ASAP!

For sure! As I said my plan is to use [Mozart](https://github.com/coenjacobs/mozart) with a couple of small changes. I'll work on that after [we start using the Composer Bin plugin](https://github.com/woocommerce/woocommerce/pull/27753).

**Comment by Konamiman** [comment]
I'm merging without approvals because @ObliviousHarmony actually approved it, but I rebased from master to fix merge conflicts and after force-pushing GitHub removed the approval. (Note to self: merge, not rebase, if the pull request has been approved already)

**Pull request merged**
